src/cargo/util/template.rs

```diff
@@ -0,0 +1,231 @@
+    }
+    Ok(())
+}
+
+/// Trait to hold information required for rendering templated files.
+pub trait TemplateFile {
+    /// Path of the template output for the file being written.
+    fn path(&self) -> &Path;
+
+    /// Return the template string.
+    fn template(&self) -> CargoResult<String>;
+}
+
+/// TemplateFile based on an input file.
+pub struct InputFileTemplateFile {
+    input_path: PathBuf,
+    output_path: PathBuf,
+}
+
+impl TemplateFile for InputFileTemplateFile {
+    fn path(&self) -> &Path {
+        &self.output_path
+    }
+
+    fn template(&self) -> CargoResult<String> {
+        let mut template_str = String::new();
+        let mut entry_file = try!(File::open(&self.input_path).chain_error(|| {
+            human(format!("Failed to open file for templating: {}", self.input_path.display()))
+        }));
+        try!(entry_file.read_to_string(&mut template_str).chain_error(|| {
+            human(format!("Failed to read file for templating: {}", self.input_path.display()))
+        }));
+        Ok(template_str)
+    }
+}
+
+impl InputFileTemplateFile {
+    pub fn new(input_path: PathBuf, output_path: PathBuf) -> InputFileTemplateFile {
+        InputFileTemplateFile {
+            input_path: input_path,
+            output_path: output_path
+        }
+    }
+}
+
+/// An in memory template file for --bin or --lib.
+pub struct InMemoryTemplateFile {
+    template_str: String,
+    output_path: PathBuf,
+}
+
+impl TemplateFile for InMemoryTemplateFile {
+    fn path(&self) -> &Path {
+        &self.output_path
+    }
+
+    fn template(&self) -> CargoResult<String> {
+        Ok(self.template_str.clone())
+    }
+}
+
+impl InMemoryTemplateFile {
+    pub fn new(output_path: PathBuf, template_str: String) -> InMemoryTemplateFile {
+        InMemoryTemplateFile {
+            template_str: template_str,
+            output_path: output_path
+        }
+    }
+}
+
+pub enum TemplateDirectory{
+    Temp(TempDir),
+    Normal(PathBuf),
+}
+
+impl TemplateDirectory {
+    pub fn path(&self) -> &Path {
+        match *self {
+            TemplateDirectory::Temp(ref tempdir) => tempdir.path(),
+            TemplateDirectory::Normal(ref path) => path.as_path()
+        }
+    }
+}
+
+/// A listing of all the files that are part of the template.
+pub struct TemplateSet {
+    pub template_dir: Option<TemplateDirectory>,
+    pub template_files: Vec<Box<TemplateFile>>
+}
+
+// The type of template we will use.
+#[derive(Debug, Eq, PartialEq)]
+pub enum TemplateType<'a>  {
+    GitRepo(&'a str),
+    LocalDir(&'a str),
+    Builtin
+}
+
+/// Given a repository string and subdir, determine if this is a git repository, local file, or a
+/// built in template.
+pub fn get_template_type<'a>(repo: Option<&'a str>,
+                             subdir: Option<&'a str>) -> CargoResult<TemplateType<'a>> {
+    match (repo, subdir) {
+        (Some(repo_url), _) if repo_url.starts_with("http://")  ||
```

Comment:
Sure. It seems there is a bug in the cargo git wrappers, though. `file://` is supported by git as a git protocol. And raw `git2` supports `file://` as a scheme. But the cargo wrappers crap out. Instead of my hack, I think it should probably just be fixed in the wrapper.